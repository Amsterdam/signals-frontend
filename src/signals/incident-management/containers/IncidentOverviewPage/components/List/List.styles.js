import styled, { css } from 'styled-components';

export const List = styled.div`
  width: 100%;

  ${({ isLoading }) =>
    isLoading &&
    css`
      opacity: 0.3;
    `}
`;

export const Table = styled.table`
  border-collapse: separate;
  width: 100%;
  height: 100%;

  td {
    padding: 0;

    a {
      text-decoration: none;
      color: black;
      display: block;
      width: 100%;
      height: 100%;
      padding: 8px;
    }
  }

  tr:hover td,
  td {
    box-shadow: unset;
  }
`;

export const Th = styled.th`
  cursor: pointer;
  font-weight: normal;
  white-space: nowrap;

  &:hover {
    text-decoration: underline;
  }
`;

export const Td = styled.td`
  ${({ noWrap }) =>
    noWrap &&
    css`
      white-space: nowrap;
    `}
`;
