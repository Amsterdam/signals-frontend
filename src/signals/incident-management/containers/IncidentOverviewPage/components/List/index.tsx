// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2018 - 2023 Gemeente Amsterdam
import type { FunctionComponent, ReactNode } from 'react'
import { useEffect, useState } from 'react'

import { Play } from '@amsterdam/asc-assets'
import differenceInCalendarDays from 'date-fns/differenceInCalendarDays'
import parseISO from 'date-fns/parseISO'
import { Link, useNavigate } from 'react-router-dom'

import ParentIncidentIcon from 'components/ParentIncidentIcon'
import configuration from 'shared/services/configuration/configuration'
import { formatAddress } from 'shared/services/format-address'
import {
  getListIconByKey,
  getListValueByKey,
} from 'shared/services/list-helpers/list-helpers'
import { string2date, string2time } from 'shared/services/string-parser'
import { statusList } from 'signals/incident-management/definitions'
import type {
  Definition,
  Priority,
  Status,
} from 'signals/incident-management/definitions/types'
import { INCIDENT_URL } from 'signals/incident-management/routes'
import type { IncidentList, IncidentListItem } from 'types/api/incident-list'
import onButtonPress from 'utils/on-button-press'

import {
  BaseTh,
  ContentSpan,
  StyledIcon,
  StyledList,
  Table,
  TdStyle,
  ThArea,
  ThDate,
  ThDay,
  ThId,
  ThParent,
  ThPriority,
  ThStatus,
  ThSubcategory,
  Tr,
} from './styles'
import { useIncidentManagementContext } from '../../../../context'
import { SortOptionLabels, SortOptions } from '../../contants'
import { compareSortOptions } from '../../utils'
import ThSort from '../Th'

export const getDaysOpen = (incident: IncidentListItem) => {
  const statusesWithoutDaysOpen = statusList
    .filter(({ shows_remaining_sla_days }) => !shows_remaining_sla_days)
    .map(({ key }) => key)
  const hasDaysOpen =
    incident.status && !statusesWithoutDaysOpen.includes(incident.status.state)

  const createdAtDate = parseISO(incident.created_at)

  if (!hasDaysOpen || isNaN(createdAtDate.getTime())) return '-'

  return differenceInCalendarDays(new Date(), createdAtDate)
}
interface TdProps {
  detailLink: string
  children: ReactNode
}

const Td = ({ detailLink, children, ...rest }: TdProps) => {
  return (
    <TdStyle {...rest}>
      <Link to={detailLink} tabIndex={-1}>
        <ContentSpan>{children}</ContentSpan>
      </Link>
    </TdStyle>
  )
}

const ChildIcon: FunctionComponent = () => (
  <StyledIcon aria-label="Deelmelding" data-testid="child-icon">
    <Play />
  </StyledIcon>
)

interface ListProps {
  className?: string
  incidents: IncidentList
  isLoading?: boolean
  priority: Priority[]
  stadsdeel: Definition[]
  status: Status[]
  orderingChangedAction: (ordering: string) => void
  ordering?: SortOptions
  sortingDisabled?: boolean
}

const List: FunctionComponent<ListProps> = ({
  className,
  incidents,
  isLoading = false,
  priority,
  stadsdeel,
  status,
  ordering,
  orderingChangedAction,
  sortingDisabled = false,
}) => {
  const { districts, referrer } = useIncidentManagementContext()
  const navigate = useNavigate()

  const [lastId, setLastId] = useState<number | undefined>()

  const navigateToIncident = (id: number) => {
    navigate(`../${INCIDENT_URL}/${id}`)
  }

  /**
   * This method reverses the ordering if the column is already ordered
   * otherwise it will order the new column ascending
   * @param column
   */
  const changeOrder = (column: SortOptions) => {
    if (ordering && compareSortOptions(ordering, column)) {
      if (ordering.startsWith('-')) {
        orderingChangedAction(ordering.replace('-', ''))
      } else {
        orderingChangedAction(`-${ordering}`)
      }
    } else {
      orderingChangedAction(column)
    }
  }

  useEffect(() => {
    // check if referrer is incident detail page
    if (referrer?.startsWith('/manage/incident/')) {
      const lastIncidentId = Number(referrer.replace('/manage/incident/', ''))
      setLastId(lastIncidentId)
    }
  }, [referrer])

  // reset lastId on click
  useEffect(() => {
    const handleClick = () => {
      setLastId(undefined)
    }

    window.addEventListener('click', handleClick)
    return () => window.removeEventListener('click', handleClick)
  }, [])

  return (
    <StyledList
      isLoading={isLoading}
      className={className}
      data-testid="incident-overview-list-component"
    >
      <Table cellSpacing="0">
        <thead>
          <tr>
            <ThParent />
            <ThSort
              StyledComponent={ThPriority}
              sortOption={SortOptions.PRIORITY_ASC}
              headerText={SortOptionLabels.PRIORITY}
              ordering={ordering}
              changeOrder={changeOrder}
              sortingDisabled={sortingDisabled}
            />
            <ThSort
              StyledComponent={ThId}
              sortOption={SortOptions.ID_DESC}
              headerText={SortOptionLabels.ID}
              ordering={ordering}
              changeOrder={changeOrder}
              sortingDisabled={sortingDisabled}
            />
            <ThDay $isDisabled={true}>Dag</ThDay>
            <ThSort
              StyledComponent={ThDate}
              sortOption={SortOptions.CREATED_AT_DESC}
              headerText={SortOptionLabels.DATE}
              ordering={ordering}
              changeOrder={changeOrder}
            />
            <ThSort
              StyledComponent={ThSubcategory}
              sortOption={SortOptions.SUBCATEGORY_ASC}
              headerText={SortOptionLabels.SUBCATEGORY}
              ordering={ordering}
              changeOrder={changeOrder}
              sortingDisabled={sortingDisabled}
            />
            <ThSort
              StyledComponent={ThStatus}
              sortOption={SortOptions.STATUS_ASC}
              headerText={SortOptionLabels.STATUS}
              ordering={ordering}
              changeOrder={changeOrder}
              sortingDisabled={sortingDisabled}
            />
            {configuration.featureFlags.fetchDistrictsFromBackend ? (
              <ThSort
                StyledComponent={ThArea}
                sortOption={SortOptions.DISTRICT_ASC}
                headerText={configuration.language.district}
                ordering={ordering}
                changeOrder={changeOrder}
                sortingDisabled={sortingDisabled}
              />
            ) : (
              <ThSort
                StyledComponent={ThArea}
                sortOption={SortOptions.BUROUGH_ASC}
                headerText={SortOptionLabels.BUROUGH}
                ordering={ordering}
                changeOrder={changeOrder}
                sortingDisabled={sortingDisabled}
              />
            )}
            <ThSort
              StyledComponent={BaseTh}
              sortOption={SortOptions.ADDRESS_ASC}
              headerText={SortOptionLabels.ADDRESS}
              ordering={ordering}
              changeOrder={changeOrder}
              sortingDisabled={sortingDisabled}
            />
            {configuration.featureFlags.assignSignalToEmployee && (
              <ThSort
                StyledComponent={BaseTh}
                sortOption={SortOptions.ASSIGNED_USER_EMAIL_ASC}
                headerText={SortOptionLabels.ASSIGNED_USER_EMAIL}
                ordering={ordering}
                changeOrder={changeOrder}
                sortingDisabled={sortingDisabled}
              />
            )}
          </tr>
        </thead>
        <tbody>
          {incidents.map((incident) => {
            const detailLink = `/manage/incident/${incident.id}`
            return (
              <Tr
                key={incident.id}
                tabIndex={0}
                onKeyDown={(e) => {
                  onButtonPress(e, () => navigateToIncident(incident.id))
                }}
                $lastIncident={incident.id === lastId}
              >
                <Td detailLink={detailLink} data-testid="incident-parent">
                  {incident.has_children && <ParentIncidentIcon />}
                  {incident.has_parent && <ChildIcon />}
                </Td>
                <Td detailLink={detailLink} data-testid="incident-urgency">
                  <StyledIcon>
                    {getListIconByKey(priority, incident.priority?.priority)}
                  </StyledIcon>
                </Td>
                <Td detailLink={detailLink} data-testid="incident-id">
                  {incident.id}
                </Td>
                <Td detailLink={detailLink} data-testid="incident-days-open">
                  {getDaysOpen(incident)}
                </Td>
                <Td detailLink={detailLink} data-testid="incident-created-at">
                  {string2date(incident.created_at)}{' '}
                  {string2time(incident.created_at)}
                </Td>
                <Td detailLink={detailLink} data-testid="incident-subcategory">
                  {incident.category?.sub}
                </Td>
                <Td detailLink={detailLink} data-testid="incident-status">
                  {getListValueByKey(status, incident.status?.state)}
                </Td>
                <Td detailLink={detailLink} data-testid="incident-area">
                  {configuration.featureFlags.fetchDistrictsFromBackend
                    ? getListValueByKey(districts, incident.location?.area_code)
                    : getListValueByKey(
                        stadsdeel,
                        incident.location?.stadsdeel
                      )}
                </Td>
                <Td detailLink={detailLink} data-testid="incident-address">
                  {incident.location?.address &&
                    formatAddress(incident.location?.address)}
                </Td>
                {configuration.featureFlags.assignSignalToEmployee && (
                  <Td
                    detailLink={detailLink}
                    data-testid="incident-assigned-user"
                  >
                    {incident.assigned_user_email}
                  </Td>
                )}
              </Tr>
            )
          })}
        </tbody>
      </Table>
    </StyledList>
  )
}

export default List
