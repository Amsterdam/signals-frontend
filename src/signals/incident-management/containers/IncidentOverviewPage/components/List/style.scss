.list-component {
  padding: 10px;

  tr:hover {
    cursor: pointer;
  }

  &__title {
    font-size: 16px;
    font-weight: bold;
    margin-bottom: 10px;
  }

}
