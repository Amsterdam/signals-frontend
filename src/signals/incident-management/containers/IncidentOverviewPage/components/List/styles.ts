// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2021 - 2023 Gemeente Amsterdam
import { themeSpacing, Icon, themeColor } from '@amsterdam/asc-ui'
import styled, { css } from 'styled-components'

export const StyledList = styled.div<{ isLoading?: boolean }>`
  width: 100%;
  overflow: auto;

  ${({ isLoading }) => isLoading && 'opacity: 0.3;'}
`

export const Table = styled.table`
  border-collapse: separate;
  width: 100%;
  height: 100%;

  tr > th > svg {
    height: ${themeSpacing(4)};
    margin-left: ${themeSpacing(2.5)};
    margin-bottom: -${themeSpacing(0.5)};
    fill: ${themeColor('primary')};
  }
`

const Th = styled.th`
  white-space: nowrap;
`

export const Tr = styled.tr<{ $lastIncident?: boolean }>`
  :focus {
    outline: auto;
  }

  &:hover td,
  td {
    box-shadow: unset;
  }

  /* Because of a quirk in Safari, the box shadow needs to be set on the cell instead of the row */
  ${({ $lastIncident }) =>
    $lastIncident &&
    css`
      & td,
      &:hover td {
        border-bottom: none;
        box-shadow: black 0px 2px 0px 0px inset, black 0px -2px 0px 0px inset;
      }

      & td:first-child,
      &:hover td:first-child {
        box-shadow: black 0px 2px 0px 0px inset, black 0px -2px 0px 0px inset,
          black 2px 0px 0px 0px inset;
      }

      & td:last-child,
      &:hover td:last-child {
        box-shadow: black 0px 2px 0px 0px inset, black 0px -2px 0px 0px inset,
          black -2px 0px 0px 0px inset;
      }
    `}
`

export const BaseTh = styled(Th)<{ $isDisabled?: boolean }>`
  cursor: pointer;
  color: ${themeColor('primary')};

  &:hover {
    text-decoration: underline;
    color: ${themeColor('primary', 'dark')};
  }

  ${({ $isDisabled }) =>
    $isDisabled &&
    css`
      &,
      &:hover {
        cursor: auto;
        text-decoration: none;
        color: ${themeColor('tint', 'level7')};
      }
    `}
`

export const ThParent = styled(Th)`
  width: 45px;
  min-height: 1px;
`

export const ThPriority = styled(BaseTh)`
  width: 32px;
`

export const ThDate = styled(BaseTh)`
  width: 150px;
`

export const ThArea = styled(BaseTh)`
  width: 120px;
`

export const ThSubcategory = styled(BaseTh)`
  width: 220px;
`

export const ThStatus = styled(BaseTh)`
  width: 160px;
`

export const ThId = styled(BaseTh)`
  width: 95px;
`

export const ThDay = styled(BaseTh)`
  width: 55px;
`

export const ContentSpan = styled.span``

export const TdStyle = styled.td`
  padding: 0;
  height: 60px;

  a {
    display: flex;
    align-items: center;
    padding: 0 ${themeSpacing(2)};
    height: 100%;

    text-decoration: none;
    color: black;

    ${ContentSpan} {
      // Show ellipsis on second line
      display: -webkit-box;
      -webkit-line-clamp: 2;
      -webkit-box-orient: vertical;
      line-clamp: 2;
      overflow: hidden;
      text-overflow: ellipsis;
      height: auto;
    }
  }
`

export const StyledIcon = styled(Icon).attrs({
  size: 16,
  role: 'img',
})`
  & svg {
    fill: ${themeColor('tint', 'level4')};
  }
`
