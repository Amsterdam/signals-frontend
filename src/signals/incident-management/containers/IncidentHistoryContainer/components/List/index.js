import React from 'react';
import PropTypes from 'prop-types';

import { string2date, string2time } from 'shared/services/string-parser/string-parser';
import './style.scss';

class List extends React.Component { // eslint-disable-line react/prefer-stateless-function
  render() {
    const { incidentHistoryList } = this.props;
    return (
      <div className="incident-history-container-list">
        <div className="incident-history-container-list__body">
          <table className="incident-history-container-list__body-list" cellSpacing="0" cellPadding="0">
            <thead>
              <tr>
                <th className="">Wanneer</th>
                <th className="">Wat</th>
                <th className="">Omschrijving</th>
                <th className="">Gebruiker</th>
              </tr>
            </thead>
            <tbody>
              {incidentHistoryList.map((item) => (
                <tr key={item.identifier}>
                  <td>{string2date(item.when)} om {string2time(item.when)}</td>
                  <td>{item.action}</td>
                  <td className="pre-wrap">{item.description}</td>
                  <td>{item.who}</td>
                </tr>
              ))
              }
            </tbody>
          </table>
        </div>
      </div>
    );
  }
}

List.propTypes = {
  incidentHistoryList: PropTypes.array.isRequired,
};

List.defaultProps = {
  incidentHistoryList: []
};

export default List;
