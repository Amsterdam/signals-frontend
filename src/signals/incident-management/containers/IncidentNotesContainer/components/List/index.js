import React from 'react';
import PropTypes from 'prop-types';

import { string2date, string2time } from 'shared/services/string-parser/string-parser';
import './style.scss';

class List extends React.Component { // eslint-disable-line react/prefer-stateless-function
  render() {
    const { incidentNotesList } = this.props;
    return (
      <div className="incident-notes-container-list">
        <div className="incident-notes-container-list__body">
          <table className="incident-notes-container-list__body-list" cellSpacing="0" cellPadding="0">
            <thead>
              <tr>
                <th className="">Datum</th>
                <th className="">Tijd</th>
                <th className="">Notitie</th>
                <th className="">Gebruiker</th>
              </tr>
            </thead>
            <tbody>
              {incidentNotesList.map((item) => (
                <tr key={item._links.self.href}>
                  <td>{string2date(item.created_at)}</td>
                  <td>{string2time(item.created_at)}</td>
                  <td className="pre-wrap">{item.text}</td>
                  <td>{item.created_by}</td>
                </tr>
              ))
              }
            </tbody>
          </table>
        </div>
      </div>
    );
  }
}

List.propTypes = {
  incidentNotesList: PropTypes.array.isRequired,
};

List.defaultProps = {
  incidentNotesList: []
};

export default List;
