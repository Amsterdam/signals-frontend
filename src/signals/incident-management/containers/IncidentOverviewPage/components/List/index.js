import React from 'react';
import PropTypes from 'prop-types';

import { string2date, string2time } from 'shared/services/string-parser/string-parser';
import { getListValueByKey } from 'shared/services/list-helper/list-helper';
import './style.scss';

class List extends React.Component { // eslint-disable-line react/prefer-stateless-function
  selectIncident = (incident) => () => {
    this.props.incidentSelected(incident);
  }

  render() {
    const { incidents, incidentsCount, statusList, stadsdeelList } = this.props;
    return (
      <div className="list-component">
        <div className="list-component__title">Meldingen ({incidentsCount})</div>

        <div className="list-component__body">
          <table className="" cellSpacing="0" cellPadding="0">
            <thead>
              <tr>
                <th className="">Id</th>
                <th className="">Datum</th>
                <th className="">Tijd</th>
                <th className="">Stadsdeel</th>
                <th className="">Rubriek</th>
                <th className="">Afdeling</th>
                <th className="">Status</th>
                <th className="">Adres</th>
              </tr>
            </thead>
            <tbody>
              {incidents.map((incident) => (
                <tr key={incident.id} onClick={this.selectIncident(incident)}>
                  <td>{incident.id}</td>
                  <td>{string2date(incident.incident_date_start)}</td>
                  <td>{string2time(incident.incident_date_start)}</td>
                  <td>{getListValueByKey(stadsdeelList, incident.location.stadsdeel)}</td>
                  <td>{incident.category.sub}</td>
                  <td>{incident.category.department}</td>
                  <td>{getListValueByKey(statusList, incident.status.state)}</td>
                  <td>{incident.location.address_text}</td>
                </tr>
              ))
              }
            </tbody>
          </table>
        </div>
      </div>
    );
  }
}

List.propTypes = {
  incidentsCount: PropTypes.number,
  incidents: PropTypes.array.isRequired,
  statusList: PropTypes.array.isRequired,
  stadsdeelList: PropTypes.array.isRequired,

  incidentSelected: PropTypes.func.isRequired,
};

export default List;
