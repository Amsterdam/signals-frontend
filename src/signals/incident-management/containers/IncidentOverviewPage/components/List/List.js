import React, { useCallback, useContext } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import parseISO from 'date-fns/parseISO';
import differenceInCalendarDays from 'date-fns/differenceInCalendarDays';
import { ChevronUp, ChevronDown } from '@datapunt/asc-assets';
import { Icon } from '@datapunt/asc-ui';

import { string2date, string2time } from 'shared/services/string-parser';
import { getListValueByKey } from 'shared/services/list-helper/list-helper';
import * as types from 'shared/types';
import configuration from 'shared/services/configuration/configuration';
import { statusList } from 'signals/incident-management/definitions';

import * as S from './List.styles';
import IncidentManagementContext from '../../../../context';

const getDaysOpen = incident => {
  const statusesWithoutDaysOpen = statusList
    .filter(({ shows_remaining_sla_days }) => shows_remaining_sla_days === false)
    .map(({ key }) => key);
  const hasDaysOpen = incident.status && !statusesWithoutDaysOpen.includes(incident.status.state);
  const start = hasDaysOpen && parseISO(incident.created_at);
  return hasDaysOpen ? -differenceInCalendarDays(start, new Date()) : '-';
};

const List = ({
  className = '',
  incidents,
  isLoading = false,
  onChangeOrdering,
  priority,
  sort,
  stadsdeel,
  status,
}) => {
  const { districts, users } = useContext(IncidentManagementContext);

  const onSort = useCallback(
    newSort => () => {
      const sortIsAsc = sort?.indexOf(newSort) === 0;
      onChangeOrdering(sortIsAsc ? `-${newSort}` : newSort);
    },
    [onChangeOrdering, sort]
  );

  const renderChevron = useCallback(
    column => {
      const currentSort = sort?.split(',')[0];
      const isColumnSorted = currentSort?.indexOf(column) > -1;
      const sortDirection = currentSort?.charAt(0) === '-' ? 'down' : 'up';

      return isColumnSorted ? (
        <Icon inline size={12}>
          {sortDirection === 'up' ? <ChevronDown /> : <ChevronUp />}
        </Icon>
      ) : null;
    },
    [sort]
  );

  const getAssignedUserName = useCallback(userId => users?.find(user => user.id === userId)?.username, [users]);

  return (
    <S.List isLoading={isLoading} className={className} data-testid="incidentOverviewListComponent">
      <S.Table cellSpacing="0">
        <thead>
          <tr>
            <S.Th data-testid="sortId" onClick={onSort('id')}>
              Id {renderChevron('id')}
            </S.Th>
            <S.Th data-testid="sortDaysOpen" onClick={onSort('days_open')}>
              Dag {renderChevron('days_open')}
            </S.Th>
            <S.Th data-testid="sortCreatedAt" onClick={onSort('created_at')}>
              Datum en tijd {renderChevron('created_at')}
            </S.Th>
            {configuration.fetchDistrictsFromBackend ? (
              <S.Th data-testid="sortDistrict" onClick={onSort('district,-created_at')}>
                {configuration.language.district} {renderChevron('district')}
              </S.Th>
            ) : (
              <S.Th data-testid="sortStadsdeel" onClick={onSort('stadsdeel,-created_at')}>
                Stadsdeel {renderChevron('stadsdeel')}
              </S.Th>
            )}
            <S.Th data-testid="sortSubcategory" onClick={onSort('sub_category,-created_at')}>
              Subcategorie {renderChevron('sub_category')}
            </S.Th>
            <S.Th data-testid="sortStatus" onClick={onSort('status,-created_at')}>
              Status {renderChevron('status')}
            </S.Th>
            <S.Th data-testid="sortPriority" onClick={onSort('priority,-created_at')}>
              Urgentie {renderChevron('priority')}
            </S.Th>
            <S.Th data-testid="sortAddress" onClick={onSort('address,-created_at')}>
              Adres {renderChevron('address')}
            </S.Th>
            {users && <th data-testid="sortAssigedUserId">Toegewezen aan</th>}
          </tr>
        </thead>
        <tbody>
          {incidents.map(incident => {
            const detailLink = `/manage/incident/${incident.id}`;
            return (
              <tr key={incident.id}>
                <S.Td>
                  <Link to={detailLink}>{incident.id}</Link>
                </S.Td>
                <S.Td data-testid="incidentDaysOpen">
                  <Link to={detailLink}>{getDaysOpen(incident)}</Link>
                </S.Td>
                <S.Td noWrap>
                  <Link to={detailLink}>
                    {string2date(incident.created_at)} {string2time(incident.created_at)}
                  </Link>
                </S.Td>
                <S.Td>
                  <Link to={detailLink}>
                    {configuration.fetchDistrictsFromBackend
                      ? getListValueByKey(districts, incident.location && incident.location.area_code)
                      : getListValueByKey(stadsdeel, incident.location && incident.location.stadsdeel)}
                  </Link>
                </S.Td>
                <S.Td>
                  <Link to={detailLink}>{incident.category && incident.category.sub}</Link>
                </S.Td>
                <S.Td>
                  <Link to={detailLink}>{getListValueByKey(status, incident.status && incident.status.state)}</Link>
                </S.Td>
                <S.Td>
                  <Link to={detailLink}>
                    {getListValueByKey(priority, incident.priority && incident.priority.priority)}
                  </Link>
                </S.Td>
                <S.Td>
                  <Link to={detailLink}>{incident.location && incident.location.address_text}</Link>
                </S.Td>
                {users && (
                  <S.Td>
                    <Link to={detailLink}>{getAssignedUserName(incident.assigned_user_id)}</Link>
                  </S.Td>
                )}
              </tr>
            );
          })}
        </tbody>
      </S.Table>
    </S.List>
  );
};

List.propTypes = {
  className: PropTypes.string,
  incidents: PropTypes.arrayOf(types.incidentType).isRequired,
  onChangeOrdering: PropTypes.func.isRequired,
  isLoading: PropTypes.bool,
  priority: types.dataListType.isRequired,
  sort: PropTypes.string,
  stadsdeel: types.dataListType.isRequired,
  status: types.dataListType.isRequired,
};

export default List;
