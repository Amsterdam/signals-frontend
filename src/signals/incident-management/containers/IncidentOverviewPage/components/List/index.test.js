import React from 'react';
import { shallow } from 'enzyme';

import List from './index';

describe('<List />', () => {
  let renderedComponent;
  let props;

  beforeEach(() => {
    props = {
      incidents: [{ id: 158, signal_id: 'bdd90784-6093-449e-bacb-bec415446b6c', source: 'api', text: 'poep', text_extra: 'extra_boten_snelheid_rondvaartboot: Ja, extra_boten_snelheid_rederij: Aemstelland', status: { id: 187, text: 'test sophie', user: 'S.Raeymaekers@amsterdam.nl', target_api: '', state: 'i', extern: false, extra_properties: { IP: '194.13.133.3' } }, location: { id: 155, stadsdeel: 'A', buurt_code: 'abc', address: { postcode: '1012JS', huisletter: 'A', huisnummer: '1', woonplaats: 'Amsterdam', openbare_ruimte: 'Dam', huisnummer_toevoeging: '1' }, address_text: 'Dam 1A-1 1012JS Amsterdam', geometrie: { type: 'Point', coordinates: [52.36227330330391, 4.930282384157181] }, extra_properties: {} }, category: { main: '', sub: 'Gezonken boot', department: '', priority: null }, reporter: { email: 'a@b.com', phone: '020654321', remove_at: '2018-07-25T13:46:10.035176Z', created_at: '2018-07-11T13:46:10.072042Z', updated_at: '2018-07-11T13:46:10.072057Z', extra_properties: null }, created_at: '2018-07-11T13:46:10.073510Z', updated_at: null, incident_date_start: '2018-07-11T13:46:10.073488Z', incident_date_end: null, operational_date: null, image: 'https://ae70d54aca324d0480ca01934240c78f.objectstore.eu/signals/images/2018/07/11/2.7mb_afbeeling_l1b9nqX.jpg?temp_url_sig=126bd0830babcf7717e44b613861c70a25c1e557&temp_url_expires=1531407297', extra_properties: null }],
      statusList: [],
      stadsdeelList: [],
      incidentSelected: jest.fn()
    };

    renderedComponent = shallow(
      <List {...props} />
    );
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should render correctly', () => {
    expect(renderedComponent).toMatchSnapshot();
  });

  it('should contain one incident', () => {
    expect(renderedComponent.find('tbody > tr').length).toEqual(1);
  });

  it('should select the incident when the row is clicked', () => {
    renderedComponent.find('tbody > tr').simulate('click');
    expect(props.incidentSelected).toHaveBeenCalledWith(props.incidents[0]);
  });
});
