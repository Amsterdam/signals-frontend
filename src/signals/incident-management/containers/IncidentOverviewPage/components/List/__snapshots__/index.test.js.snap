// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> should render correctly 1`] = `
<div
  className="list-component"
>
  <div
    className="list-component__title"
  >
    Meldingen (
    )
  </div>
  <div
    className="list-component__body"
  >
    <table
      cellPadding="0"
      cellSpacing="0"
      className=""
    >
      <thead>
        <tr>
          <th
            className=""
          >
            Id
          </th>
          <th
            className=""
          >
            Datum
          </th>
          <th
            className=""
          >
            Tijd
          </th>
          <th
            className=""
          >
            Stadsdeel
          </th>
          <th
            className=""
          >
            Rubriek
          </th>
          <th
            className=""
          >
            Afdeling
          </th>
          <th
            className=""
          >
            Status
          </th>
          <th
            className=""
          >
            Adres
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          key="158"
          onClick={[Function]}
        >
          <td>
            158
          </td>
          <td>
            11-07-2018
          </td>
          <td>
            15:46
          </td>
          <td>
            Niet gevonden
          </td>
          <td>
            Gezonken boot
          </td>
          <td />
          <td>
            Niet gevonden
          </td>
          <td>
            Dam 1A-1 1012JS Amsterdam
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
